File: modules/services.js
// modules/services.js
const { exec } = require('child_process');
const util = require('util');
const execPromise = util.promisify(exec);

class ServiceManager {
    constructor(bot) {
        this.bot = bot;
    }

    // Получить статус конкретной службы
    async getServiceStatus(serviceName) {
        // Сначала пробуем без sudo
        try {
            const { stdout } = await execPromise(`systemctl status ${serviceName} --no-pager -n 5`);
            const activeMatch = stdout.match(/Active: (\w+)/);
            const status = activeMatch ? activeMatch[1] : 'unknown';
            
            // Парсим дополнительную информацию
            const loadMatch = stdout.match(/Loaded: (.+?)\n/);
            const loaded = loadMatch ? loadMatch[1] : 'unknown';
            
            const pidMatch = stdout.match(/Main PID: (\d+)/);
            const pid = pidMatch ? pidMatch[1] : null;
            
            const memoryMatch = stdout.match(/Memory: ([\d.]+[KMG])/i);
            const memory = memoryMatch ? memoryMatch[1] : null;
            
            return {
                name: serviceName,
                status,
                loaded,
                pid,
                memory,
                details: stdout.split('\n').slice(-5).join('\n')
            };
        } catch (error) {
            // Если не получилось без sudo, пробуем с sudo
            try {
                const sudoPassword = process.env.SUDO_PASSWORD;
                let stdout;
                
                if (sudoPassword) {
                    // Используем пароль из переменной окружения
                    const result = await execPromise(
                        `echo '${sudoPassword}' | sudo -S systemctl status ${serviceName} --no-pager -n 5`
                    );
                    stdout = result.stdout;
                } else {
                    // Пробуем без пароля
                    const result = await execPromise(`sudo systemctl status ${serviceName} --no-pager -n 5`);
                    stdout = result.stdout;
                }
                
                const activeMatch = stdout.match(/Active: (\w+)/);
                const status = activeMatch ? activeMatch[1] : 'unknown';
                
                const loadMatch = stdout.match(/Loaded: (.+?)\n/);
                const loaded = loadMatch ? loadMatch[1] : 'unknown';
                
                const pidMatch = stdout.match(/Main PID: (\d+)/);
                const pid = pidMatch ? pidMatch[1] : null;
                
                const memoryMatch = stdout.match(/Memory: ([\d.]+[KMG])/i);
                const memory = memoryMatch ? memoryMatch[1] : null;
                
                return {
                    name: serviceName,
                    status,
                    loaded,
                    pid,
                    memory,
                    details: stdout.split('\n').slice(-5).join('\n')
                };
            } catch (sudoError) {
                return {
                    name: serviceName,
                    status: 'inactive',
                    details: error.stdout || error.message || sudoError.message
                };
            }
        }
    }

    // Управление службой (start/stop/restart)
    async controlService(serviceName, action) {
        // Сначала пробуем без sudo
        try {
            const { stdout } = await execPromise(`systemctl ${action} ${serviceName}`);
            return { success: true, message: stdout || `Служба ${action} выполнена` };
        } catch (error) {
            // Если не получилось без sudo, пробуем с sudo
            try {
                // Проверяем, есть ли пароль в переменной окружения
                const sudoPassword = process.env.SUDO_PASSWORD;
                
                if (sudoPassword) {
                    // Используем echo для передачи пароля в sudo через -S
                    const { stdout } = await execPromise(
                        `echo '${sudoPassword}' | sudo -S systemctl ${action} ${serviceName}`
                    );
                    return { success: true, message: stdout || `Служба ${action} выполнена` };
                } else {
                    // Пробуем без пароля (если настроен NOPASSWD)
                    const { stdout } = await execPromise(`sudo systemctl ${action} ${serviceName}`);
                    return { success: true, message: stdout || `Служба ${action} выполнена` };
                }
            } catch (sudoError) {
                // Проверяем, требует ли sudo пароль
                if (sudoError.message && sudoError.message.includes('password')) {
                    return { 
                        success: false, 
                        message: 'Требуется настройка sudo. Варианты:\n\n' +
                                '1. Установите переменную окружения:\n' +
                                '   export SUDO_PASSWORD="ваш_пароль"\n\n' +
                                '2. Или настройте sudo без пароля:\n' +
                                '   sudo visudo\n' +
                                '   Добавьте: YOUR_USER ALL=(ALL) NOPASSWD: /bin/systemctl\n\n' +
                                '3. Или запустите бота от root пользователя.'
                    };
                }
                return { success: false, message: sudoError.message || error.message };
            }
        }
    }

    // Получить логи службы
    async getServiceLogs(serviceName, lines = 20) {
        // Сначала пробуем без sudo
        try {
            const { stdout } = await execPromise(`journalctl -u ${serviceName} -n ${lines} --no-pager`);
            return stdout;
        } catch (error) {
            // Если не получилось без sudo, пробуем с sudo
            try {
                const sudoPassword = process.env.SUDO_PASSWORD;
                let stdout;
                
                if (sudoPassword) {
                    // Используем пароль из переменной окружения
                    const result = await execPromise(
                        `echo '${sudoPassword}' | sudo -S journalctl -u ${serviceName} -n ${lines} --no-pager`
                    );
                    stdout = result.stdout;
                } else {
                    // Пробуем без пароля
                    const result = await execPromise(`sudo journalctl -u ${serviceName} -n ${lines} --no-pager`);
                    stdout = result.stdout;
                }
                
                return stdout;
            } catch (sudoError) {
                return `❌ Ошибка получения логов: ${sudoError.message || error.message}`;
            }
        }
    }
}

module.exports = ServiceManager;

File: internal/services/services.go
//go:build linux

package services

import (
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"strings"
)

func GetServiceStatus(serviceName string) (*ServiceStatus, error) {
	tryCmds := [][]string{
		{"systemctl", "status", serviceName, "--no-pager", "-n", "5"},
	}
	if os.Getenv("SUDO_PASSWORD") != "" {
		tryCmds = append(tryCmds, []string{"sh", "-c", "echo '" + os.Getenv("SUDO_PASSWORD") + "' | sudo -S systemctl status " + serviceName + " --no-pager -n 5"})
	}
	tryCmds = append(tryCmds, []string{"sudo", "systemctl", "status", serviceName, "--no-pager", "-n", "5"})

	for _, args := range tryCmds {
		cmd := exec.Command(args[0], args[1:]...)
		out, err := cmd.CombinedOutput()
		if err != nil {
			continue
		}
		return parseSystemctlStatus(string(out), serviceName), nil
	}
	return &ServiceStatus{
		Name:   serviceName,
		Status: "inactive",
	}, nil
}

func parseSystemctlStatus(stdout, serviceName string) *ServiceStatus {
	s := &ServiceStatus{Name: serviceName}
	activeRe := regexp.MustCompile(`Active: (\w+)`)
	if m := activeRe.FindStringSubmatch(stdout); len(m) > 1 {
		s.Status = m[1]
	}
	loadRe := regexp.MustCompile(`Loaded: (.+?)\n`)
	if m := loadRe.FindStringSubmatch(stdout); len(m) > 1 {
		s.Loaded = strings.TrimSpace(m[1])
	}
	pidRe := regexp.MustCompile(`Main PID: (\d+)`)
	if m := pidRe.FindStringSubmatch(stdout); len(m) > 1 {
		s.PID = m[1]
	}
	memRe := regexp.MustCompile(`Memory: ([\d.]+[KMG])`)
	if m := memRe.FindStringSubmatch(stdout); len(m) > 1 {
		s.Memory = m[1]
	}
	lines := strings.Split(stdout, "\n")
	if len(lines) > 5 {
		s.Details = strings.Join(lines[len(lines)-5:], "\n")
	}
	return s
}

func ControlService(serviceName, action string) *ControlResult {
	tryCmds := [][]string{
		{"systemctl", action, serviceName},
	}
	if os.Getenv("SUDO_PASSWORD") != "" {
		tryCmds = append(tryCmds, []string{"sh", "-c", "echo '" + os.Getenv("SUDO_PASSWORD") + "' | sudo -S systemctl " + action + " " + serviceName})
	}
	tryCmds = append(tryCmds, []string{"sudo", "systemctl", action, serviceName})

	for _, args := range tryCmds {
		cmd := exec.Command(args[0], args[1:]...)
		out, err := cmd.CombinedOutput()
		if err != nil {
			if strings.Contains(err.Error(), "password") || strings.Contains(string(out), "password") {
				return &ControlResult{
					Success: false,
					Message: "Требуется настройка sudo. Варианты:\n\n1. export SUDO_PASSWORD=\"ваш_пароль\"\n2. sudo visudo → NOPASSWD: /bin/systemctl\n3. Запустите от root",
				}
			}
			continue
		}
		msg := strings.TrimSpace(string(out))
		if msg == "" {
			msg = "Служба " + action + " выполнена"
		}
		return &ControlResult{Success: true, Message: msg}
	}
	return &ControlResult{Success: false, Message: "Не удалось выполнить " + action}
}

func GetServiceLogs(serviceName string, lines int) string {
	lineArg := fmt.Sprintf("%d", lines)
	tryCmds := [][]string{
		{"journalctl", "-u", serviceName, "-n", lineArg, "--no-pager"},
	}
	if os.Getenv("SUDO_PASSWORD") != "" {
		tryCmds = append(tryCmds, []string{"sh", "-c", "echo '" + os.Getenv("SUDO_PASSWORD") + "' | sudo -S journalctl -u " + serviceName + " -n " + lineArg + " --no-pager"})
	}
	tryCmds = append(tryCmds, []string{"sudo", "journalctl", "-u", serviceName, "-n", lineArg, "--no-pager"})

	for _, args := range tryCmds {
		cmd := exec.Command(args[0], args[1:]...)
		out, err := cmd.CombinedOutput()
		if err != nil {
			continue
		}
		return string(out)
	}
	return "❌ Ошибка получения логов"
}
